services/worker-manager/test/provider_azure_test.js

```diff
@@ -0,0 +1,377 @@
+      workerId: 'whatever',
+      providerId,
+      created: taskcluster.fromNow('-1 hour'),
+      expires: taskcluster.fromNow('1 week'),
+      capacity: 1,
+      lastModified: taskcluster.fromNow('-2 weeks'),
+      lastChecked: taskcluster.fromNow('-2 weeks'),
+      state: helper.Worker.states.REQUESTED,
+      providerData: {
+        ...baseProviderData,
+        registrationExpiry: Date.now() + 1000,
+      },
+    });
+    await provider.scanPrepare();
+    await provider.checkWorker({worker});
+    await provider.scanCleanup();
+    assert(!fakeAzure.deleteVMStub.called);
+    assert(!fakeAzure.deleteDiskStub.called);
+    assert(!fakeAzure.deleteIPStub.called);
+    assert(!fakeAzure.deleteNICStub.called);
+  });
+
+  suite('registerWorker', function() {
+    const workerGroup = providerId;
+    const workerId = '5d06deb3-807b-46dd-aef5-78aaf9193f71';
+
+    const defaultWorker = {
+      workerPoolId,
+      workerGroup,
+      workerId,
+      providerId,
+      created: taskcluster.fromNow('0 seconds'),
+      lastModified: taskcluster.fromNow('0 seconds'),
+      lastChecked: taskcluster.fromNow('0 seconds'),
+      capacity: 1,
+      expires: taskcluster.fromNow('90 seconds'),
+      state: 'requested',
+      providerData: {
+        ...baseProviderData,
+      },
+    };
+
+    test('document is not a valid PKCS#7 message', async function() {
+      const worker = await helper.Worker.create({
+        ...defaultWorker,
+      });
+      const document = 'this is not a valid PKCS#7 message';
+      const workerIdentityProof = {document};
+      await assert.rejects(() =>
+        provider.registerWorker({workerPool, worker, workerIdentityProof}),
+      /Signature validation error/);
+      assert(monitorManager.messages[0].Fields.error.includes('Too few bytes to read ASN.1 value.'));
+    });
+
+    test('document is empty', async function() {
+      const worker = await helper.Worker.create({
+        ...defaultWorker,
+      });
+      const document = '';
+      const workerIdentityProof = {document};
+      await assert.rejects(() =>
+        provider.registerWorker({workerPool, worker, workerIdentityProof}),
+      /Signature validation error/);
+      assert(monitorManager.messages[0].Fields.error.includes('Too few bytes to parse DER.'));
+    });
+
+    test('message does not match signature', async function() {
+      const worker = await helper.Worker.create({
+        ...defaultWorker,
+      });
+      // this file is a version of `azure_signature_good` where vmId has been edited in the message
+      const document = fs.readFileSync(path.resolve(__dirname, 'fixtures/azure_message_bad')).toString();
+      const workerIdentityProof = {document};
+      await assert.rejects(() =>
+        provider.registerWorker({workerPool, worker, workerIdentityProof}),
+      /Signature validation error/);
+      assert(monitorManager.messages[0].Fields.error.includes('Only 8, 16, 24, or 32 bits supported: 264'));
```

Comment:
This and the next test are passing but seem to fail with a failure that is not testing the thing we intended? I need to figure out  more about `'Only 8, 16, 24, or 32 bits supported: 264'` before we ship this I think

Reply:
I think this is because we just arbitarily futzed with a valid message to get an invalid one. We probably have to actually encode a real message with an invalid cert to get this test to be acccurate. Does that seem true, @milescrabill?

I'm pretty sure this is _working_ right now but just that the test is not testing it.